core/src/main/java/com/uber/m3/tally/ScopeImpl.java

```diff
@@ -42,14 +43,14 @@
     private ScheduledExecutorService scheduler;
     private Registry registry;
 
-    // ConcurrentHashMap nearly always allowing read operations seems like a good
-    // performance upside to the consequence of reporting a newly-made metric in
-    // the middle of looping and reporting through all metrics. Therefore, we only
-    // synchronize on these maps when having to allocate new metrics.
-    private final Map<String, CounterImpl> counters = new ConcurrentHashMap<>();
-    private final Map<String, GaugeImpl> gauges = new ConcurrentHashMap<>();
-    private final Map<String, TimerImpl> timers = new ConcurrentHashMap<>();
-    private final Map<String, HistogramImpl> histograms = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, CounterImpl> counters = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, GaugeImpl> gauges = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, HistogramImpl> histograms = new ConcurrentHashMap<>();
+
+    // TODO validate if needs to be concurrent at all
+    private final ConcurrentLinkedQueue<MetricBase> reportingQueue = new ConcurrentLinkedQueue<>();
```

Comment:
Tally client already has heavy memory footprint, what's the memory impact of the additional queue?
Also how much cpu reduction does switching from traversing the hashmaps to traversing the queue? It is also not clear to me why that would reduce cpu usage.

Reply:
To reduce the traversal scope to only those metrics that had actually changed. This is leveraged in #74.